compiler/cpp/src/thrift/generate/t_csharp_generator.cc

```diff
@@ -1254,10 +1254,10 @@ void t_csharp_generator::generate_csharp_union_definition(std::ostream& out, t_s
   indent_up();
 
   indent(out) << "public abstract void Write(TProtocol protocol);" << endl;
-  indent(out) << "public readonly bool Isset;" << endl;
-  indent(out) << "public abstract object Data { get; }" << endl;
+  indent(out) << "public readonly int Isset;" << endl;
+  indent(out) << "protected abstract object Data { get; }" << endl;
 
-  indent(out) << "protected " << tunion->get_name() << "(bool isset) {" << endl;
+  indent(out) << "protected " << tunion->get_name() << "(int isset) {" << endl;
```

Comment:
Is this change from bool to int an optimization?

Reply:
Depends how you define that … it stores the information which member is set **exactly**. The type alone may not do this:

    union message_piece {
      1: Participant Sender
      2: Participant Receiver
      3. string  Contents
    }

The type alone is not sufficient here, because trying to cast into ```Participant``` alone will not tell you, which case you actually have. 

I admit that the example is a bit artificial, but the construct as such is neither illegal nor absolutely unexpected. We use union constructs to wrap messages sent back and forth using that model, because it makes certain things easier to handle for us. 

Long story short: Yes, there is a good reason behind.